fdbserver/BlobManifest.actor.cpp

```diff
@@ -0,0 +1,369 @@
+		state Reference<BackupContainerFileSystem> readBstore = self->blobConn_->getForRead(MANIFEST_FILENAME);
+		state Reference<IAsyncFile> reader = wait(readBstore->readFile(MANIFEST_FILENAME));
+		state int64_t fileSize = wait(reader->size());
+		state Arena arena;
+		state uint8_t* data = new (arena) uint8_t[fileSize];
+		int readSize = wait(reader->read(data, fileSize, 0));
+		dprint("Blob manifest restoring {} bytes\n", readSize);
+		StringRef ref = StringRef(data, readSize);
+		return Value(ref, arena);
+	}
+
+	// Decode blob manifest from binary data
+	static Standalone<BlobManifest> decode(Value data) {
+		Standalone<BlobManifest> manifest;
+		BinaryReader binaryReader(data, IncludeVersion());
+		binaryReader >> manifest;
+		return manifest;
+	}
+
+	// Write system keys to database
+	ACTOR static Future<Void> writeSystemKeys(Reference<BlobManifestLoader> self, VectorRef<KeyValueRef> rows) {
+		state Transaction tr(self->db_);
+		loop {
+			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+			tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+			try {
+				for (auto& row : rows) {
+					tr.set(row.key, row.value);
+				}
+				wait(tr.commit());
+				dprint("Blob manifest loaded {} rows\n", rows.size());
+				return Void();
+			} catch (Error& e) {
+				wait(tr.onError(e));
+			}
+		}
+	}
+
+	// Iterate active granules and return their version/sizes
+	ACTOR static Future<BlobGranuleVersionVector> listGranules(Reference<BlobManifestLoader> self) {
+		state Transaction tr(self->db_);
+		loop {
+			state BlobGranuleVersionVector results;
+			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+			tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+
+			try {
+				std::vector<KeyRangeRef> granules;
+				state int i = 0;
+				auto limit = GetRangeLimits::BYTE_LIMIT_UNLIMITED;
+				state RangeResult blobRanges = wait(tr.getRange(blobGranuleMappingKeys, limit));
+				for (i = 0; i < blobRanges.size() - 1; i++) {
+					Key startKey = blobRanges[i].key.removePrefix(blobGranuleMappingKeys.begin);
+					Key endKey = blobRanges[i + 1].key.removePrefix(blobGranuleMappingKeys.begin);
+					state KeyRange granuleRange = KeyRangeRef(startKey, endKey);
+					try {
+						Standalone<BlobGranuleVerion> granule = wait(getGranule(&tr, granuleRange));
+						results.push_back_deep(results.arena(), granule);
+					} catch (Error& e) {
+						dprint("missing data for key range {} \n", granuleRange.toString());
+					}
+				}
+				return results;
+			} catch (Error& e) {
+				wait(tr.onError(e));
+			}
+		}
+	}
+
+	// Find the newest granule for a key range. The newest granule has the max version and relevant files
+	ACTOR static Future<Standalone<BlobGranuleVerion>> getGranule(Transaction* tr, KeyRangeRef range) {
+		state Standalone<BlobGranuleVerion> granuleVersion;
```

Comment:
Why do you not want to use getLatestGranuleHistory()? Because it might not have files?
It would still be better to use a backwards scan implementation than a forwards scan here though, I'd think

Reply:
Yes. Latest granule history may not have granule files. I need to find the last granule history with files. 

Updated to use a forward scan. it seems more straightforward. thanks!